File: 155-min-stack/min-stack.cpp
class MinStack {
private:
    std::stack<int> _stk;
    std::stack<int> _min;
public:
    MinStack() {}
    
    void push(int val) {
        if (_min.empty() || _min.top() >= val) {
            _min.push(val);
        }
        _stk.push(val);
    }
    
    void pop() {
        if (_min.top() == _stk.top()) {
            _min.pop();
        }
        _stk.pop();
    }
    
    int top() {
        return _stk.top();
    }
    
    int getMin() {
        return _min.top();
    }
};

File: 155-min-stack/min-stack.py
class MinStack:

    def __init__(self):
        self._stack_val = []
        self._stack_min = []

    def push(self, val: int) -> None:
        self._stack_val.append(val)
        self._stack_min.append(val if not self._stack_min else min(val, self._stack_min[-1]))

    def pop(self) -> None:
        self._stack_val.pop()
        self._stack_min.pop()

    def top(self) -> int:
        return self._stack_val[-1] if self._stack_val else -1

    def getMin(self) -> int:
        return self._stack_min[-1] if self._stack_min else -1


# Your MinStack object will be instantiated and called as such:
# obj = MinStack()
# obj.push(val)
# obj.pop()
# param_3 = obj.top()
# param_4 = obj.getMin()